### DISCUSSION TITLE
AppGroups - Allow vs Block

### DISCUSSION THREAD

**Discussion by ghost** [tnodir/fort#141] [Ideas]
There should be a way to deal with this:

For example, I added `C:\Program Files\WindowsApps\**` In the Block column and then `C:\Program Files\WindowsApps\Microsoft.WindowsStore_*_x64__8wekyb3d8bbwe\WinStore.App.exe` in the Allow.

It says it is being blocked, because Block takes priority over Allow.

Well, that's the conclusion in my simple test of adding the same .exe in both columns and it always got blocked.

I was thinking maybe there should be a way to set a priority for each group and one group will take priority? So I can add `WinStore.App.exe` in one group and then in a 'low priority' the whole folder, but then, it seems a little too much for something so rare to use.

I don't even know if it possible with how Fort Firewall logic is done for the AppGroups. or if there is a way to set Allow to be priority and not Block inside the AppGroups.

To explain the reason for it:
I am using Programs as a way to see if some apps want to connect to the internet and that way to allow or block them. The problem with using `Ignore, if not blocked or allowed` is that apps that are not allowed or blocked don't even appear in the Statistics -> Blocked Connections, which makes sense. But also disabling `Collect New Programs` will not let me see apps in Programs.

So using `Block, if not allowed` would add them to Programs/Logger, allowing me to see what is being blocked and the other one will half do what I want with it.

I guess I wanted a workaround about FF not having a true notification system for blocked apps, but even with a notification system, there would be the same issue, unless there was a way to ignore specific folders from being 'logged' by Fort, like Ignore mode works but for specific folders.

So maybe there should be a way to exclude folders and work like Ignore mode where `Any new apps will be ignored and will not appear in the 'Programs' window.` but inside `Block, if not allowed`?

Maybe like a way to change the Application Group's column from "block" to "Ignore" or a way to show an `ignore` AppGroup column or something like that to take advantage of wildcards and that way folders there will not be logged, or blocked but just ignored.
Mainly meant for folders but if people want to use it for individual files it will work like if an Application Group had the `Collect blocked connections` off, so technically there will be no difference.

I think this still would be good for the future, because if there was a notification system and Fort asked people if you want to allow or block a connection, it would be good to have a way to ignore certain folders and exes from it, so something that works for both Programs and Application Groups would be good, or just Application Groups.

Of course this idea depends on how AppGroups does its things and how the Block Notification system works in the future, maybe there is no way to work around it and only block everything manually each program is the only option, but with the possibility of using wildcards.

But I even tested wildcard `!` but I don't think it is supported in Fort and documentation says:
>An optional prefix "!" which negates the pattern; any matching file excluded by a previous pattern will become included again. It is not possible to re-include a file if a parent directory of that file is excluded.
Would means it won't work in this case.

For now I am blocking them individually but thought I would throw an idea about it and feedback in case anything I said it is useful for the future.

**Comment by tnodir** [comment]
> I added C:\Program Files\WindowsApps\** In the Block column and then C:\Program Files\WindowsApps\Microsoft.WindowsStore_*_x64__8wekyb3d8bbwe\WinStore.App.exe in the Allow.

> It says it is being blocked, because Block takes priority over Allow.

Well, actually the logic is simpler:
- all wildcard paths are collected to one alphabetically sorted list
- and driver processes the wildcard paths one by one from start

So, the `C:\Program Files\WindowsApps\**` path precedes the `C:\Program Files\WindowsApps\Microsoft.WindowsStore_*_x64__8wekyb3d8bbwe\WinStore.App.exe`.

**Reply by ghost** [reply]
Yeah but I added `C:\program files\windowsapps\microsoft.windowsstore_22310.1401.8.0_x64__8wekyb3d8bbwe\winstore.app.exe` (or any program) to both columns and it got blocked, like the block takes priority over Allow, even if technically everything is 'blocked' by the mode.
That made me think the logic was like: Since everything is blocked inside `WindowsApps` and subfolders, then Allowing takes no place there.

I mean, why would anyone have the same path in both columns is beyond me lol, unless someone accidentally made many groups and did it accidentally, but in that case the Allow will not work unless the Block gets removed, I don't know if that was meant to be.

But since we are talking about wildcard here, and `\**` vs some exe inside a subfolder there, hope it can be improved 👍would be good to block all Downloads for example and only allow few exe from it, which means the 'allow' doesn't necessarily means it will have a wildcard.

**Comment by tnodir** [comment]
Let me describe my plan to improve it in the #142.

**Reply by tnodir** [reply]
> But I even tested wildcard ! but I don't think it is supported in Fort

Yes, the `!` is not supported.

The wildcards processing just exits after 1-st matched path.

**Comment by ghost** [comment]
Posting here because it is kind of topic - off-topic

> > At least Groups in Programs, makes more sense than Options, especially if the approach is to make them work more together to 'fix' these kind of situations.
> 
> I like this idea. Added #143.


Oh nice.

BTW @tnodir
Wouldn't `QTabBar::RoundedNorth` look better than `QTabBar::TriangularNorth`?
I will have to guess the use of Triangular was to make them look different? but Triangular look out of place with the whole nice rounded Fort QT UI, especially in dark mode where a ugly white border is drawn against the gray/dark UI.

Fort even supports the mica win11 effect in the titlebar, something still some apps today don't do, like browsers (unless experimental), but I don't know if QT will add mica anywhere else anytime soon.
I still remember the older UI, it was nice with the squared tabs and the black background, it looked nice, but I guess QT makes it easier especially with the dark mode they finally support, of course, the white border in the tabs ruins it though, maybe without it it wouldn't look so bad, but the shape doesn't seem to match Fort UI at all.

![image](https://github.com/tnodir/fort/assets/152097351/0e25e380-4c17-4468-9d88-748a6b98899b)

Also, only 3 places use it AppGroups, Addresses and Statistics. 
https://github.com/tnodir/fort/blob/de1ad7533b5d9b4cec45ed9114843807aeb0776c/src/ui/form/stat/pages/trafficpage.cpp#L216

https://github.com/tnodir/fort/blob/de1ad7533b5d9b4cec45ed9114843807aeb0776c/src/ui/form/opt/pages/applicationspage.cpp#L316

https://github.com/tnodir/fort/blob/de1ad7533b5d9b4cec45ed9114843807aeb0776c/src/ui/form/opt/pages/addressespage.cpp#L101

It's been so many years with the triangular ones, so I better throw the suggestion.

Anyway, I am posting old UI that I found on the internet just for the nostalgia lol.

<img src=https://pesktop.com/oO3yT94ENRsWHxb38XPeClgZwuQuQZawRabjobBvCM8/w:700/h:400/rt:fill/el:1/wm:0.8:soea:16:16:0.2/aHR0cHM6Ly9wZXNrdG9wLmNvbS91cGxvYWRzLzE3ZTJjMWZhNzczY2ZiY2FhN2IzZTRlOTA1Yzg4OTM1LndlYnA.jpg >

**Reply by tnodir** [reply]
> Wouldn't QTabBar::RoundedNorth look better than QTabBar::TriangularNorth?

![fort-tab](https://github.com/tnodir/fort/assets/303655/d0720dbe-4392-402e-a5f0-4888823dfa91)

What do you think?

**Reply by tnodir** [reply]
> Fort even supports the mica win11 effect in the titlebar

Qt supports it. I use latest Qt from `dev` branch.

**Reply by ghost** [reply]
Tabs look much better for sure, maybe close icon could be improved to match it better. but it looks better 100% 👍Triangular just doesn't look nice in Fort for sure.
